odbc/src/api/functions.rs

```diff
@@ -1279,73 +1274,218 @@ pub unsafe extern "C" fn SQLFetch(statement_handle: HStmt) -> SqlReturn {
+    let mut row_error_count = 0;
+
+    // Use `index` to figure out which buffer in the array of buffers to use.
+    for index in 0..rowset_size {
+        let move_to_next_result = {
+            let connection = must_be_valid!((*stmt.connection).as_connection());
+            match stmt.mongo_statement.write().unwrap().as_mut() {
+                Some(mongo_stmt) => mongo_stmt
+                    .next(connection.mongo_connection.read().unwrap().as_ref())
+                    .map_err(|e| e.into()),
+                None => Err(ODBCError::InvalidCursorState),
+            }
+        };
+
+        // if row_status_ptr is null, it was not set by the user and can be ignored.
+        // Otherwise, we need to set the row_status_buffer to the address that corresponds with the current row.
+        let row_status_buffer: *mut USmallInt =
+            if !stmt.attributes.read().unwrap().row_status_ptr.is_null() {
+                (stmt.attributes.read().unwrap().row_status_ptr as ULen
+                    + (index * size_of::<u16>())) as *mut USmallInt
+            } else {
+                null_mut()
+            };
 
-                match sql_return {
-                    SqlReturn::ERROR => return SqlReturn::ERROR,
-                    SqlReturn::SUCCESS_WITH_INFO => {
-                        success_with_info_encountered = true;
+        if let Ok((has_next, mut warnings_opt)) = move_to_next_result {
+            warnings_opt.iter().for_each(|warning| {
+                add_diag_with_function!(
+                    MongoHandleRef::from(statement_handle),
+                    ODBCError::GeneralWarning(warning.to_string()),
+                    function_name.to_string()
+                );
+            });
+
+            if !has_next {
+                // There are no more rows, so set the rest of Row Status Array buffers to SQL_ROW_NOROW if the rowset end has not been reached.
+                if !row_status_buffer.is_null() {
+                    for row_status_index in index..rowset_size {
+                        let row_status_buffer = (stmt.attributes.read().unwrap().row_status_ptr
+                            as ULen
+                            + (row_status_index * size_of::<u16>()))
+                            as *mut USmallInt;
+                        *row_status_buffer = RowStatus::SQL_ROW_NOROW as USmallInt;
                     }
-                    _ => {}
+                }
+
+                // break here to prevent NO_DATA from being returned if the rowset had at least one row in it.
+                if has_fetched_at_least_one_row {
+                    break;
+                }
+
+                // if the cursor has already reached the end of the result set when SQLFetch is called, NO_DATA is returned.
+                // Additionally, rows_fetch_ptr is set to 0.
+                return SqlReturn::NO_DATA;
+            }
+
+            has_fetched_at_least_one_row = true;
+
+            if !row_status_buffer.is_null() {
```

Comment:
If you use a marker, here you can do 
```
if track_row_status {
   let row_status_buffer: *mut USmallInt = (stmt.attributes.read().unwrap().row_status_ptr as ULen + (index * size_of::<u16>())) as *mut USmallInt;
    *row_status_buffer = if warnings_opt.is_empty() {
        RowStatus::SQL_ROW_SUCCESS as USmallInt
    } else {
        RowStatus::SQL_ROW_SUCCESS_WITH_INFO as USmallInt
    };
}
```

Reply:
I made a marker called `has_row_status_array`, but is it really preferable to define `row_status_buffer` within every `if has_row_status_array {...}` block rather than just defining it once like I am already doing. (i.e., 
```
        let row_status_buffer: *mut USmallInt =
            if has_row_status_array {
                (stmt.attributes.read().unwrap().row_status_ptr as ULen
                    + (index * size_of::<u16>())) as *mut USmallInt
            } else {
                null_mut()
            };
```
).
I'm open to doing it either way, but the current way avoids code duplication.